### DISCUSSION TITLE
Heterogeneous Graph Classification with empty node types

### DISCUSSION THREAD

**Discussion by DogitoErgoSum** [pyg-team/pytorch_geometric#4281] [Q&A] [answered]
Hi,
I am intending to perform heterogeneous graph classification. Each graph mostly consists of empty node types as shown below:

![image](https://user-images.githubusercontent.com/51344466/158676969-8d2eb306-6ff3-4816-a1e4-9b890bcb9ef8.png)

All the edges are directed since the node features happen at different times.
Right now, I have this toy model that I was able to build piece by piece since my code abilities and understanding is very limited

`from torch_geometric.nn import HeteroConv, GraphConv,GATv2Conv
from torch_geometric.nn import GATConv, Linear, to_hetero
class HeteroGNN(torch.nn.Module):
def __init__(self, metadata, hidden_channels, out_channels, num_layers,heads):
        super().__init__()

        self.convs = torch.nn.ModuleList()
        for _ in range(num_layers):
            conv = HeteroConv({edge_type: GATv2Conv(-1, hidden_channels,heads) for edge_type in data.edge_types
                          }, aggr='sum')
            self.convs.append(conv)

        self.lin = Linear(hidden_channels, out_channels)

    def forward(self, x_dict, edge_index_dict):
        for conv in self.convs:
            x_dict = conv(x_dict, edge_index_dict)
            x_dict = {key: F.sigmoid(x) for key, x in x_dict.items()}
        return x_dict
modelo=HeteroGNN(data.metadata(),10,5,2,5).to("cuda:0")` 

I applied this model to a single graph and obtained the following:

`import torch.nn.functional as F
modelo(data.x_dict,data.edge_index_dict)`

![image](https://user-images.githubusercontent.com/51344466/158677949-df9261de-dbad-43f4-b13e-8a39630b5307.png)

How can I aggregate information about the present nodes, ignoring the empty nodes, in order to make a graph classification?

**Comment by DogitoErgoSum** [comment]
i guess i will add a virtual node as solution for now. Any other solution pls write

**Comment by rusty1s** [comment, marked as answer]
Any reason you want to maintain at all? Is this for batching reasons only?

Your output looks correct to me. You now only need to aggregate the information from node types that are present:
```python
out = 0
for node_type in x_dict.keys():
    if x_dict[node_type].numel() == 0:
        continue
    out += global_mean_pool(x_dict[node_type], batch_dict[node_type])
```
Let me know if this works for you!

**Reply by DogitoErgoSum** [reply]
Yes the empty nodes are for batching reasons. Ty, I will try this and give posterior feedback

**Reply by DogitoErgoSum** [reply]
something like this?
![image](https://user-images.githubusercontent.com/51344466/158881695-3f3c412e-1d7e-48cb-b1a0-5c6abc15b7c0.png)

I get the following error:
![image](https://user-images.githubusercontent.com/51344466/158883380-9c5e207f-1159-490a-979c-4e02c9c28da0.png)
it works for a few nodes and the global mean pool shape changes
![image](https://user-images.githubusercontent.com/51344466/158883561-8afa5139-7398-4f8d-a784-9ddd950e1f63.png)

Data is a HeteroDataBatch with 3 graphs

**Reply by DogitoErgoSum** [reply]
This is works fine if i use a batch size of 1

**Reply by DogitoErgoSum** [reply]
for another batch size:
![image](https://user-images.githubusercontent.com/51344466/158892832-348188d0-e722-4cf7-bdc4-7df2cdc9c6da.png)

the batches sizes dict have different shapes

**Reply by DogitoErgoSum** [reply]
Sorry for the spam, I just added size=batch_size to global_mean_pooling and it solved the problem. TY once again

**Reply by rusty1s** [reply]
Yeah, just wanted to recommend this. Glad you resolved it already.
